### DISCUSSION TITLE
How to get the element components of an rst file?

### DISCUSSION THREAD

**Discussion by krame123** [pyansys/pymapdl-reader#315] [Q&A] [answered]
Hello everyone,I've been having some problems with pyansys recently. I'm using the rst.element_components command to output an empty collection. May I ask how to solve this problem?

**Comment by RobPasMue** [comment]
Hi @krame123! May I ask, which library are you using? Is it PyMAPDL's [``element_components``](https://mapdl.docs.pyansys.com/version/stable/api/_autosummary/ansys.mapdl.core.mesh_grpc.MeshGrpc.element_components.html#element-components)?

**Reply by krame123** [reply]
Yes, I'm using PyMAPDL's elelment_components.

**Reply by RobPasMue** [reply]
Perfect, let me redirect the discussion to the PyMAPDL repository. Pinging @germa89 and @clatapie for visibility

**Reply by germa89** [reply]
`results.element_components` or `mesh.element_components` relies on the PyMAPDL-Reader class `RST`. However see this comment for an anwer: https://github.com/ansys/pymapdl/discussions/2298#discussioncomment-6911827

**Comment by clatapie** [comment]
Hello @krame123!
Could you provide us a minimal working example of your code to have a better understanding of the issue you're encountering please?

**Reply by krame123** [reply]
Hello, thank you for your help. Since I am only new to pyansys, I only made some attempts to read the rst file, the code is as follows:
```py
from ansys.mapdl import reader as pymapdl_reader
result = pymapdl_reader.read_binary("D:/rst/file.rst")
print('------element_stress--------')
elem,estress,enode = result.element_stress(0)
print(elem)
print(elem[0])
print(enode)
print('------time-------')
time = result.time_values
print(time)
print('------components--------')
com = result.element_components
print(com)
```
In the code, I find that the output component is an empty collection. How should I change the code. In addition, I found that the output of the element_stress command is eight stress tensors, my understanding is the stress tensor at the element node, what command should I use if I want to output the stress tensor of the element.

**Reply by germa89** [reply]
Hi @krame123 

`element_components` are not implemented in [PyMAPDL-Reader](https://github.com/ansys/pymapdl-reader/blob/86819560e67bd2d210f92052337c08f3bd961715/ansys/mapdl/reader/mesh.py#L113). 
Hence this is a PyMAPDL-Reader question.

Alternatively, you could use PyMAPDL but you will need to have the capability of launch an MAPDL instance (and hence having a license to do that). You will also have to use the dev branch of the repo, since the component manager hasn't been released yet.
To install the dev branch follow [this link](https://mapdl.docs.pyansys.com/version/dev/getting_started/index.html#python-module) using the part that starts with *"For a local development version, install with:"*

If you can launch an MAPDL instance, then the procedure will be something like:

```py
from ansys.mapdl.core import launch_mapdl

mapdl = launch_mapdl()

# Read in the results file
mapdl.file("D:/rst/file.rst")

# Inspect the components in the live MAPDL instance
print(mapdl.components)
```

Further [information about the component manager](https://mapdl.docs.pyansys.com/version/dev/api/_autosummary/ansys.mapdl.core.component.ComponentManager.html#ansys.mapdl.core.component.ComponentManager)

**Reply by mikerife** [reply, marked as answer]
Hi @krame123 @germa89 @RobPasMue @clatapie 
The PyMAPDL Reader does have a method of showing the MAPDL 'components' (or WB Mechanical Named Selections):

![image](https://github.com/ansys/pymapdl/assets/91209185/0f64e96f-315d-47ca-b3ce-9758b73319e2)


However components were not always saved to the MAPDL result file.  @krame123 if you print the result, as shown in the screen capture, then what Version is given?  It's also possible that there are no components in the rst file; are you sure some were defined prior to solving?

Mike

**Reply by krame123** [reply]
Thaks for your help. My version is 22.1. As you said above, "However components were not always saved to the MAPDL result file." I created two different Geometry in workbench and assigned different material properties. So how do I read the elements in the RST file that belong to these two Geometry in the PyMAPDL Reader.

**Reply by krame123** [reply]
In addition, I found that the output of the `element_stress` command is eight stress tensors, my understanding is the stress tensor at the element node, what command should I use if I want to output the stress tensor of the element. The code is as follows:
```
elem,estress,enode = result.element_stress(0)
print(elem)
print(elem[0])
print(enode)
```

**Reply by mikerife** [reply]
Hi @krame123 
So for the first issue; create body named selections in WB Mechanical prior to solving.  These will be written as MAPDL element components.  

For the second issue, that is not a feature of PyMAPDL Reader as element results (one result per element) are calculated as needed in MAPDL.  So there is nothing for the PyMAPDL Reader to read on the result file.  If these types of results are necessary I suggest either moving to PyMAPDL as then we can work with element tables.  Or try out PyDPF-Core...or PyDPF-Post.

**Reply by germa89** [reply]
@mikerife  you are totally right.. I wasn't aware that `element_components` was working.

But again this is PyMAPDL-Reader question. I'm re-transferring.
